jobs/integration/validation.py

```diff
@@ -2205,32 +2206,82 @@ async def cleanup():
     await cleanup()
 
 
+@dataclass
+class NagiosAlerts:
+    alerts: Mapping[str, bs_ResultSet]
+    matchers: Mapping[str, Callable[[dict], bool]]
+
+    def __bool__(self):
+        """Successful if there are some alerts and all the matchers are True."""
+        if not self.alerts:
+            return False
+        return all(
+            self.matchers[severity](alerts) for severity, alerts in self.alerts.items()
+        )
+
+    def __getitem__(self, item: str) -> bs_ResultSet:
+        return self.alerts[item]
+
+    def __str__(self) -> str:
+        return ", ".join(
+            f"{severity}/{link.string}"
+            for severity, alerts in self.alerts.items()
+            for alert in alerts
+            for link in alert.find_all("a", recursive=False)
+        )
+
+
 @dataclass
 class NagiosApi:
     open: Callable
     url: str
 
-    def find_alerts(self, *severities):
+    async def find_alerts(self, **severities) -> NagiosAlerts:
         classes = {
             "critical": "statusBGCRITICAL",
             "pending": "statusPENDING",
             "ok": "statusOK",
         }
         url_data = self.open(self.url)
         soup = bs(url_data.read(), "html.parser")
-        return {
+        alerts = {
             severity: soup.find_all("td", class_=classes[severity.lower()])
             for severity in severities
         }
+        return NagiosAlerts(alerts, severities)
 
-    async def wait_for_alerts(self, **severities):
-        while True:
-            all_alerts = self.find_alerts(*severities.keys())
-            if all(
-                severities[severity](alerts) for severity, alerts in all_alerts.items()
-            ):
-                return all_alerts
-            await asyncio.sleep(5)
+    async def critical_alerts_by_app(self, *apps: str):
+        alerts = await self.find_alerts(critical=lambda c: len(c) > 0)
+        criticals = alerts["critical"]
+
+        if criticals:
+            found = {app: [] for app in apps}
+            for c in criticals:
+                for link in c.find_all("a", recursive=False):
+                    for app in apps:
+                        if app in link.string:
+                            found[app].append(link.string)
+            if all(app_crits for app_crits in found.values()):
+                log(f"Found critical errors in {', '.join(apps)}:")
+                for app, app_crits in found.items():
+                    for crit in app_crits:
+                        log(f"{app} - {crit}")
+                return True
+        return False
```

Comment:
Adding this method that looks for `critical` nagios alerts for a given set of apps.

It only returns `True` when every app given has at least one critical error.   Its just the generic version of what the test was doing in a `while True` loop [down here](https://github.com/charmed-kubernetes/jenkins/pull/1047/files#diff-a3e9600e4339acd8390a865b32d04515eba72f1929856af1e778e36d7e550fb6L2309-L2336)